### DISCUSSION TITLE
Copilot keeps importing unwanted modules.

### DISCUSSION THREAD

**Discussion by albas99** [community/community#32895] [Copilot Conversations]
I've been having issues with Github Copilot importing stuff I don't need randomly into my code. Sometimes it might just be a method I'm using from a different library or package I'm using and Copilot would just import another package that I don't need into my code and sometimes even cause buggy behavior. How can I make it to stop doing this?

**Comment by iamcjwilliams** [comment]
Hello @albas99 Our engineering team is looking into this issue

**Comment by tilmann1** [comment]
@albas99 Can you please share more details on the issues that you're seeing? Ideally concrete examples where you see suboptimal completions being generated.

**Comment by albas99** [comment]
@tilmann1 It's not about suboptimal completion being generated, it's about copilot importing packages/libraries I don't want/need. This happens when for example I type some method or function `example.method()` or `function(example)` and there's an existing package or library with that same method or function name then copilot just decides that I'm referencing stuff from that package/library(which 99.99% of the time I'm not) and just imports it for me at the top of the script.

**Reply by tilmann1** [reply]
I see, which Copilot extension are you using?

**Reply by albas99** [reply]
VScode, latest version.

**Comment by OdinIversen** [comment]
I also have this issue all the time, suprised this has not gotten more attention.

**Comment by mdnahas** [comment]
I have this issue too.  This is my #1 complaint about Copilot.  (I'm enjoying it, otherwise!)   I'm also using VS Code.

My most recent example: I have a function called subscribeCallback() and Copilot imported a random library with a function named "subscribe".  I probably typed "subs..." and hit tab and it completed the word "subscribe" _and_ added an import at the top of the file.  

The frustrating aspect of this is the change is at the top of the file.  I don't see the change.  Later, when my code doesn't run or does run properly, I'm looking at the code that I changed and not at the top of the file.  

@iamcjwilliams One possible fix is to change Copilot, but another possible fix is to improve the user interface: making sure VS Code notifies the user of a change outside the viewable area.

**Comment by fcFn** [comment]
That is an extremely annoying issue that I think only started happening recently. If fixing it so that somehow Copilot could be more accurate when inserting imports is not possible, I would really like to have an option to disable inserting import statements completely so instead I could rely on code actions only. It is tiresome to check the top of the file for rubbish imports even if you changed just one line with Copilot.

**Comment by polku** [comment]
Same here, I'd rather have Pycharm show me a big red error when I forgot the import than copilot adding a random one and discovering it when my pipeline fails 5 minutes later.

**Comment by joaopedrobiu6** [comment]
Any update on this front? I'm enjoying Copilot but this is making me disable it more and more often.

**Comment by luke-ingram-sgr** [comment]
Crazy behavior from co-pilot on VSCode, "Oh you wrote merge in a string? I'm gonna add `from heappq import merge` at the top of your python imports and not tell you...". This is bizarre and seems to be happening all the time now. It just flat out guesses imports that don't exist and were not asked for.

**Reply by luke-ingram-sgr** [reply]
More examples: Writing `parse` can make copilot import non-existent modules containing the word parse. Simply typing L at one point made copilot add `from re import L` at the top of the script.

**Comment by popiol** [comment]
This issue is still there, is there anything going on with it? It was reported in 2022 and nothing happened???

**Comment by shaunakv1** [comment]
Any update?

**Comment by nricks** [comment]
this seems to be related to this as well #8375 and nothing has been done I'm losing so much time  having to remove the random imports like:

```python
from cryptography.hazmat.asn1.asn1 import U
```

This was literally 3 minutes ago.
